File: src/bot/cogs/materials.py
"""Materials information commands cog."""

import discord
from discord.ext import commands

from models import get_db
from services.database import MaterialService, UserService
from services.sustainability import SustainabilityAnalyzer
from utils.logger import get_logger

logger = get_logger(__name__)


class MaterialsCommands(commands.Cog):
    """Commands for material information."""

    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.analyzer = SustainabilityAnalyzer()

    @commands.command(name="impact", aliases=["material"])
    async def material_impact(self, ctx: commands.Context, *, material_name: str):
        """
        Get environmental impact of a material.

        Usage: !impact <material_name>
        Example: !impact cotton
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                # Track user
                user = UserService.get_or_create(db, ctx.author.id, ctx.author.name)
                user.add_points(5)
                UserService.update(db, user)

                # Find material
                material = MaterialService.get_by_name(db, material_name)

                if not material:
                    await ctx.send(f"❌ Material '{material_name}' not found.")
                    return

                # Calculate scores
                overall_score = material.calculate_overall_score()
                grade = material.get_grade()
                category = self.analyzer.get_impact_category(overall_score)
                recommendation = self.analyzer.get_material_recommendation(material)

                # Create embed
                embed = discord.Embed(
                    title=f"🧵 {material.name}",
                    description=material.description or "No description available",
                    color=discord.Color.green() if overall_score >= 60 else discord.Color.orange()
                )

                embed.add_field(
                    name="Overall Score",
                    value=f"**{overall_score:.1f}/100** (Grade: {grade})",
                    inline=False
                )

                embed.add_field(
                    name="Type",
                    value=material.material_type.value.replace("_", " ").title(),
                    inline=True
                )

                # Environmental scores
                embed.add_field(
                    name="💧 Water Usage",
                    value=f"{material.water_usage_score:.1f}/100",
                    inline=True
                )
                embed.add_field(
                    name="🌍 Carbon Footprint",
                    value=f"{material.carbon_footprint_score:.1f}/100",
                    inline=True
                )
                embed.add_field(
                    name="♻️ Biodegradability",
                    value=f"{material.biodegradability_score:.1f}/100",
                    inline=True
                )
                embed.add_field(
                    name="🧪 Chemical Usage",
                    value=f"{material.chemical_usage_score:.1f}/100",
                    inline=True
                )
                embed.add_field(
                    name="⚡ Energy Consumption",
                    value=f"{material.energy_consumption_score:.1f}/100",
                    inline=True
                )

                # Production metrics
                if material.water_liters_per_kg:
                    embed.add_field(
                        name="Water per kg",
                        value=f"{material.water_liters_per_kg:.0f} L",
                        inline=True
                    )
                if material.co2_kg_per_kg:
                    embed.add_field(
                        name="CO2 per kg",
                        value=f"{material.co2_kg_per_kg:.2f} kg",
                        inline=True
                    )

                # Additional properties
                if material.is_biodegradable:
                    embed.add_field(
                        name="Biodegradable",
                        value=material.is_biodegradable,
                        inline=True
                    )
                if material.recycling_potential:
                    embed.add_field(
                        name="Recycling Potential",
                        value=material.recycling_potential,
                        inline=True
                    )

                # Recommendation
                embed.add_field(
                    name="Recommendation",
                    value=recommendation,
                    inline=False
                )

                embed.set_footer(text=f"Requested by {ctx.author.name} | +5 points")

                await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in impact command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred while fetching material data.")
            finally:
                db.close()

    @commands.command(name="compare")
    async def compare_materials(
        self, ctx: commands.Context, material1: str, material2: str
    ):
        """
        Compare two materials.

        Usage: !compare <material1> <material2>
        Example: !compare cotton polyester
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                # Track user
                user = UserService.get_or_create(db, ctx.author.id, ctx.author.name)
                user.add_points(7)
                UserService.update(db, user)

                # Find materials
                mat1 = MaterialService.get_by_name(db, material1)
                mat2 = MaterialService.get_by_name(db, material2)

                if not mat1:
                    await ctx.send(f"❌ Material '{material1}' not found.")
                    return
                if not mat2:
                    await ctx.send(f"❌ Material '{material2}' not found.")
                    return

                # Compare
                comparison = self.analyzer.compare_materials(mat1, mat2)

                # Create embed
                embed = discord.Embed(
                    title=f"⚖️ {mat1.name} vs {mat2.name}",
                    description="Material Comparison",
                    color=discord.Color.blue()
                )

                # Overall scores
                score1 = comparison["overall_scores"][mat1.name]
                score2 = comparison["overall_scores"][mat2.name]
                winner = mat1.name if score1 > score2 else mat2.name

                embed.add_field(
                    name="Overall Scores",
                    value=f"**{mat1.name}**: {score1:.1f}/100\n**{mat2.name}**: {score2:.1f}/100\n\n🏆 **Winner**: {winner}",
                    inline=False
                )

                # Category comparisons
                categories = [
                    ("💧 Water Usage", "water_usage"),
                    ("🌍 Carbon Footprint", "carbon_footprint"),
                    ("♻️ Biodegradability", "biodegradability")
                ]

                for emoji_name, key in categories:
                    data = comparison[key]
                    embed.add_field(
                        name=emoji_name,
                        value=f"{mat1.name}: {data[mat1.name]:.1f}\n{mat2.name}: {data[mat2.name]:.1f}\n✓ {data['better']}",
                        inline=True
                    )

                embed.set_footer(text=f"Requested by {ctx.author.name} | +7 points")

                await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in compare command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred while comparing materials.")
            finally:
                db.close()

    @commands.command(name="search")
    async def search(self, ctx: commands.Context, *, query: str):
        """
        Search for garments or materials.

        Usage: !search <query>
        Example: !search organic cotton
        """
        async with ctx.typing():
            db = next(get_db())
            try:
                from services.database import GarmentService

                # Search both garments and materials
                materials = MaterialService.search(db, query)
                garments = GarmentService.search(db, query)

                if not materials and not garments:
                    await ctx.send(f"❌ No results found for '{query}'.")
                    return

                embed = discord.Embed(
                    title=f"🔍 Search Results for '{query}'",
                    color=discord.Color.blue()
                )

                if materials:
                    materials_text = "\n".join(
                        f"• {m.name} ({m.material_type.value})"
                        for m in materials[:5]
                    )
                    embed.add_field(
                        name=f"Materials ({len(materials)})",
                        value=materials_text,
                        inline=False
                    )

                if garments:
                    garments_text = "\n".join(
                        f"• {g.name} ({g.category})"
                        for g in garments[:5]
                    )
                    embed.add_field(
                        name=f"Garments ({len(garments)})",
                        value=garments_text,
                        inline=False
                    )

                embed.set_footer(text=f"Requested by {ctx.author.name}")

                await ctx.send(embed=embed)

            except Exception as e:
                logger.error(f"Error in search command: {e}", exc_info=True)
                await ctx.send("❌ An error occurred during search.")
            finally:
                db.close()


async def setup(bot: commands.Bot):
    """Set up the cog."""
    await bot.add_cog(MaterialsCommands(bot))

File: src/commands/materials.rs
// SPDX-License-Identifier: MPL-2.0
//! Materials cog: `impact`, `compare`, `search`.

use poise::serenity_prelude as serenity;

use super::{colour, say, send_embed};
use crate::models::MaterialType;
use crate::services::{GarmentService, MaterialService, UserService};
use crate::sustainability as analyzer;
use crate::{typing, Context, Error};

async fn award(ctx: &Context<'_>, points: i64) -> Result<(), Error> {
    let db = &ctx.data().db;
    let mut user =
        UserService::get_or_create(db, ctx.author().id.get() as i64, &ctx.author().name).await?;
    user.add_points(points);
    UserService::update(db, &user).await?;
    Ok(())
}

/// Get environmental impact of a material.
#[poise::command(prefix_command, aliases("material"))]
pub async fn impact(
    ctx: Context<'_>,
    #[rest]
    #[description = "Material name"]
    material_name: String,
) -> Result<(), Error> {
    typing(&ctx).await;
    award(&ctx, 5).await?;
    let db = &ctx.data().db;

    let material = match MaterialService::get_by_name(db, &material_name).await? {
        Some(m) => m,
        None => return say(&ctx, format!("❌ Material '{material_name}' not found.")).await,
    };

    let overall = material.calculate_overall_score();
    let grade = material.get_grade();
    // NB: the original Python computed an impact "category" here but never
    // rendered it in the embed; omitted to keep output identical.
    let recommendation = analyzer::get_material_recommendation(&material);
    let type_display = material.type_display();

    let mut embed = serenity::CreateEmbed::new()
        .title(format!("🧵 {}", material.name))
        .description(
            material
                .description
                .clone()
                .unwrap_or_else(|| "No description available".into()),
        )
        .colour(if overall >= 60.0 {
            colour::green()
        } else {
            colour::orange()
        })
        .field(
            "Overall Score",
            format!("**{overall:.1}/100** (Grade: {grade})"),
            false,
        )
        .field("Type", type_display, true)
        .field(
            "💧 Water Usage",
            format!("{:.1}/100", material.water_usage_score),
            true,
        )
        .field(
            "🌍 Carbon Footprint",
            format!("{:.1}/100", material.carbon_footprint_score),
            true,
        )
        .field(
            "♻️ Biodegradability",
            format!("{:.1}/100", material.biodegradability_score),
            true,
        )
        .field(
            "🧪 Chemical Usage",
            format!("{:.1}/100", material.chemical_usage_score),
            true,
        )
        .field(
            "⚡ Energy Consumption",
            format!("{:.1}/100", material.energy_consumption_score),
            true,
        );

    if let Some(w) = material.water_liters_per_kg {
        embed = embed.field("Water per kg", format!("{w:.0} L"), true);
    }
    if let Some(c) = material.co2_kg_per_kg {
        embed = embed.field("CO2 per kg", format!("{c:.2} kg"), true);
    }
    if let Some(b) = &material.is_biodegradable {
        embed = embed.field("Biodegradable", b, true);
    }
    if let Some(r) = &material.recycling_potential {
        embed = embed.field("Recycling Potential", r, true);
    }
    embed = embed.field("Recommendation", recommendation, false);
    embed = embed.footer(serenity::CreateEmbedFooter::new(format!(
        "Requested by {} | +5 points",
        ctx.author().name
    )));
    send_embed(&ctx, embed).await
}

/// Compare two materials.
#[poise::command(prefix_command)]
pub async fn compare(
    ctx: Context<'_>,
    #[description = "First material"] material1: String,
    #[description = "Second material"] material2: String,
) -> Result<(), Error> {
    typing(&ctx).await;
    award(&ctx, 7).await?;
    let db = &ctx.data().db;

    let mat1 = match MaterialService::get_by_name(db, &material1).await? {
        Some(m) => m,
        None => return say(&ctx, format!("❌ Material '{material1}' not found.")).await,
    };
    let mat2 = match MaterialService::get_by_name(db, &material2).await? {
        Some(m) => m,
        None => return say(&ctx, format!("❌ Material '{material2}' not found.")).await,
    };

    let cmp = analyzer::compare_materials(&mat1, &mat2);
    let winner = if cmp.overall1 > cmp.overall2 {
        &cmp.material1
    } else {
        &cmp.material2
    };

    let mut embed = serenity::CreateEmbed::new()
        .title(format!("⚖️ {} vs {}", mat1.name, mat2.name))
        .description("Material Comparison")
        .colour(colour::blue())
        .field(
            "Overall Scores",
            format!(
                "**{}**: {:.1}/100\n**{}**: {:.1}/100\n\n🏆 **Winner**: {}",
                mat1.name, cmp.overall1, mat2.name, cmp.overall2, winner
            ),
            false,
        );

    for (label, m) in [
        ("💧 Water Usage", &cmp.water_usage),
        ("🌍 Carbon Footprint", &cmp.carbon_footprint),
        ("♻️ Biodegradability", &cmp.biodegradability),
    ] {
        embed = embed.field(
            label,
            format!(
                "{}: {:.1}\n{}: {:.1}\n✓ {}",
                mat1.name, m.a, mat2.name, m.b, m.better
            ),
            true,
        );
    }
    embed = embed.footer(serenity::CreateEmbedFooter::new(format!(
        "Requested by {} | +7 points",
        ctx.author().name
    )));
    send_embed(&ctx, embed).await
}

/// Search for garments or materials.
#[poise::command(prefix_command)]
pub async fn search(
    ctx: Context<'_>,
    #[rest]
    #[description = "Search query"]
    query: String,
) -> Result<(), Error> {
    typing(&ctx).await;
    let db = &ctx.data().db;

    let materials = MaterialService::search(db, &query).await?;
    let garments = GarmentService::search(db, &query).await?;

    if materials.is_empty() && garments.is_empty() {
        return say(&ctx, format!("❌ No results found for '{query}'.")).await;
    }

    let mut embed = serenity::CreateEmbed::new()
        .title(format!("🔍 Search Results for '{query}'"))
        .colour(colour::blue());

    if !materials.is_empty() {
        let text = materials
            .iter()
            .take(5)
            .map(|m| {
                let ty = MaterialType::parse(&m.material_type)
                    .map(|t| t.value())
                    .unwrap_or(m.material_type.as_str());
                format!("• {} ({})", m.name, ty)
            })
            .collect::<Vec<_>>()
            .join("\n");
        embed = embed.field(format!("Materials ({})", materials.len()), text, false);
    }
    if !garments.is_empty() {
        let text = garments
            .iter()
            .take(5)
            .map(|g| format!("• {} ({})", g.name, g.category))
            .collect::<Vec<_>>()
            .join("\n");
        embed = embed.field(format!("Garments ({})", garments.len()), text, false);
    }
    embed = embed.footer(serenity::CreateEmbedFooter::new(format!(
        "Requested by {}",
        ctx.author().name
    )));
    send_embed(&ctx, embed).await
}
